### ISSUE TITLE
Use __cdecl instead of WINAPI for user mode exports on Windows

### ISSUE THREAD

**Issue by ThadHouse** [microsoft/msquic#376]
When working with interop scenarios, such as writing C# are Python wrappers, having a common calling convention across all platforms makes writing such interop code much easier. On linux, macOS, and Windows x64 there is only 1 calling convention, and most languages represent this using `__cdecl`. On Windows 32 bit, changing the calling conventions actually makes a difference. Some runtimes will automatically use `__cdecl` if another option is used, other runtimes will actually error if a different calling convention is used. By making windows use `__cdecl`, this becomes a non issue, as all platforms just then become `__cdecl`, and everything works. Otherwise, either separate binaries per platform of managed code needs to be provided, or runtime switching between platforms need to occur to select the right calling convention.

In addition, this can even affect C scenarios using `LoadLibrary` and `GetProcAddress`. With `__cdecl`, the exported function name is just the name of the function. However with `WINAPI`, the exported function name is `_functionName@parameterStackSize`. This would be the name that would need to be passed into `GetProcAddress`, and not just `functionName`.

In terms of generated code, technically choosing `__cdecl` means the caller has to clean up the stack, which means extra code. However, in execution, extra stack location is allocated anyway, meaning cleanup of the stack already has to occur, and the only thing that changes is an increase to the size of the stack that is cleaned up at the end of a function.

There are downsides to this in some cases. For languages that choose a default calling convention, it will usually choose `WINAPI` on windows, meaning an explicit calling convention must be passed. However not all runtimes do this, and I'd argue that interop code should be a bit more explicit in what it needs, and wrappers can easily be written so end users wouldn't need to understand the distinction. Also, `__cdecl` is the default for a library in Visual Studio, so many people building libraries already know to explicitly set the calling convention.

### Proposed solution

Change

https://github.com/microsoft/msquic/blob/master/src/inc/msquic_winuser.h#L44

`#define QUIC_API                        WINAPI`

to

`#define QUIC_API                        __cdecl`

### Additional context

This would be a breaking change to the API, so this would be something that would need to be done before 1.0 is released, or it likely couldn't happen in the future.

**Comment by nibanks** [comment]
Thanks @ThadHouse for opening the issue. We had no real reason to use `WINAPI` in the first place, and just deferred to what other Windows APIs were already doing. @jkotalik any comment here as far as .NET Core using MsQuic if we change to `__cdecl`?

**Comment by jkotalik** [comment]
`__cdecl` should be fine.
